consensus/hotstuff/integration/liveness_test.go

```diff
@@ -211,8 +213,101 @@ func Test1TimeoutOutof5Instances(t *testing.T) {
+	for i := 1; i < healthyReplicas; i++ {
+		assert.Equal(t, ref.forks.FinalizedBlock(), instances[i].forks.FinalizedBlock(), "instance %d should have same finalized block as first instance")
+		assert.Equal(t, finalizedViews, FinalizedViews(instances[i]), "instance %d should have same finalized view as first instance")
+	}
+}
+
+// TestBlockDelayIsHigherThanTimeout tests an edge case protocol edge case, where
+//   - The block arrives in time for replicas to vote.
+//   - The next primary does not respond in time with a follow-up proposal,
+//     so nodes start sending TimeoutObjects.
+//   - However, eventually, the next primary successfully constructs a QC and a new
+//     block before a TC leads to the round timing out.
+//
+// This test verifies that nodes still make progress on the happy path (QC constructed),
+// despite already having initiated the timeout.
+// Example scenarios, how this timing edge case could manifest:
+//   - block delay is very close (or larger) than round duration
+//   - delayed message transmission (specifically votes) within network
+//   - overwhelmed / slowed-down primary
+//   - byzantine primary
+//
+// Implementation:
+//   - We have 4 nodes in total where the TimeoutObjects from two of them are always
+//     discarded. Therefore, no TC can be constructed.
+//   - To force nodes to initiate the timeout (i.e. send TimeoutObjects), we set
+//     the `blockRateDelay` to _twice_ the PaceMaker Timeout. Furthermore, we configure
+//     the PaceMaker to only increase timeout duration after 6 successive round failures.
+func TestBlockDelayIsHigherThanTimeout(t *testing.T) {
+	healthyReplicas := 2
+	replicasNotGeneratingTimeouts := 2
+	finalView := uint64(20)
+
+	// generate the 4 hotstuff participants
+	participants := unittest.IdentityListFixture(healthyReplicas + replicasNotGeneratingTimeouts)
+	instances := make([]*Instance, 0, healthyReplicas+replicasNotGeneratingTimeouts)
+	root := DefaultRoot()
+	// set block rate delay to be bigger than minimal timeout
+	timeouts, err := timeout.NewConfig(pmTimeout, pmTimeout, 1.5, happyPathMaxRoundFailures, pmTimeout*2)
+	require.NoError(t, err)
+
+	// set up 2 instances that fully work (incl. sending TimeoutObjects)
+	for n := 0; n < healthyReplicas; n++ {
+		in := NewInstance(t,
+			WithRoot(root),
+			WithParticipants(participants),
+			WithLocalID(participants[n].NodeID),
+			WithTimeouts(timeouts),
+			WithStopCondition(ViewFinalized(finalView)),
+		)
+		instances = append(instances, in)
+	}
+
+	// set up two instances which don't generate timeout objects
+	for n := healthyReplicas; n < healthyReplicas+replicasNotGeneratingTimeouts; n++ {
+		in := NewInstance(t,
+			WithRoot(root),
+			WithParticipants(participants),
+			WithLocalID(participants[n].NodeID),
+			WithTimeouts(timeouts),
+			WithStopCondition(ViewFinalized(finalView)),
+			WithOutgoingTimeoutObjects(BlockAllTimeoutObjects),
```

Comment:
I'm still not convinced [this case](https://github.com/onflow/flow-go/pull/3201#discussion_r978120788) can't happen: 

> > We have 4 nodes in total where the TimeoutObjects from two of them are always discarded. Therefore, no TC can be constructed.
> 
> I'm not sure about this. 2/4 nodes will block outgoing TOs (`droppingReplicas`). However these nodes will still store their own TOs locally and may receive TOs from other healthy nodes. A `droppingReplica` could store its TO, then receive the TOs from both healthy nodes, and then be able to produce TC.
> 
> I think we also need to block inbound TOs on the two `droppingReplica`s.

Since we only block outbound TOs, a replica in `replicasNotGeneratingTimeouts` could:
* generate a TO locally, and store it
* receive 2 other TOs from the 2 nodes in `healthyReplicas`
* create a timeout certificate

I'm worried this test might be flaky, if the above condition can happen sometimes. If we block incoming TOs on the two `replicasNotGeneratingTimeouts`, then it's impossible.